tests: benchmark: Add support for nrf series processors

youvedeep opened the pull request:
time_info tests have been designed for loapic and cortex m systic
timer. nrf platforms which are arm based but use rtc timer.

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

nashif commented on tests/benchmarks/timing_info/testcase.yaml:
```diff
@@ -1,4 +1,5 @@
 tests:
 -   test:
         arch_whitelist: x86 arm
+        platform_exclude: 96b_carbon_nrf51 nrf51_blenano nrf51_pca10028 nrf51_vbluno51 nrf52840_pca10056 nrf52_blenano2 nrf52_pca10040 nrf52_vbluno52
```
use a filter instead:

filter: not CONFIG_NRF_RTC_TIMER

galak reviewed (changes requested):
We should fix the test, this either shows a flaw in what we are doing with timers or the test if we can't write code that uses timers but is agnostic to the HW that implements that timer.

youvedeep commented:
I was trying to enable this benchmark test for nrf based devices. 

For nrf51, it seems to be working . I am not getting any hang.
For nrf52, I am getting hang 

**When MMU is Enabled**
tc_start() - Time Measurement
Timing Results: Clock Frequency: 0 MHz
Context switch                              ***** MPU FAULT *****
  Executing thread ID (thread): 0x2000110c
  Faulting instruction address:  0x200024a8
  Instruction Access Violation
Fatal fault in ISR! Spinning...

**When I disabled MMU** , I am getting following error:-
tc_start() - Time tc_start() - Time Measurement
Timing Results: Clock Frequency: 0 MHz
Context switch                              ***** USAGE FAULT *****
  Executing thread ID (thread): 0x200010b8
  Faulting instruction address:  0x20002450
  Illegal use of the EPSR
Fatal fault in ISR! Spinning...

As both NRF51 and NRF52 use RTC timers and test is working on nrf51 and it is getting failed on nrf52.
One of difference between 2 is MMU support is present on nrf52. But even after disabling MMU it is getting failed. @galak  Can you please suggest what may be going on crazy on NRF52 when MMU is disabled ?

galak commented:
@carlescufi would be better person to ask.

carlescufi commented:
@youvedeep honestly I am clueless as to why this is happening. There is no real difference in the RTC hardware between nRF51 and nRF52, the differences are mostly in the core (Cortex-M0 vs Cortex-M4) and additional peripherals in the nRF52. This shoulds like an issue in either the nRF52 Zephyr port in general or (more likely) in the nRF RTC timer driver implementation that only appears when running on nRF52, perhaps because of a race condition that manifests itself only when we execute too fast. Is there a way you can get a backtrace for us? Is it easy to reproduce, just by running the test on nRF52?

carlescufi commented:
@youvedeep @galak the problem with this test is that the whole ARM benchmarking is architectured around the SysTick peripheral. The nRF51 does not have a SysTick, whereas the nRF52 has one (but one that stops whenever the IC goes to sleep). This is probably what causes the crash on nRF52, which I can reproduce here locally. What we need is a kernel dev (or potentially us when we have time) to send patches so that the ARM benchmarking code both in tests/ but also in arch/arm is able to use a non-Systick timer, so this requires a JIRA/GitHub issue. In the meantime I agree with disabling this test.

carlescufi commented:
@youvedeep any chance you can join #zephyrproject on IRC so we can chat a bit when you have the occasion?

youvedeep commented:
Thanks @carlescufi for the comments.

 
I did following experiments around this:
1. We are able to reproduce issue even with power management disabled, not sure how IC is going to sleep. Does it enters into sleep Automatically ? Also  "Low Power states/ Constant latency" modes are currently not enabled on nrf. So looks like this is not the issue.
2. I modified the code, and was calling nrf RTC timer based APIs in case of nrf board. something like :
+#if defined(CONFIG_CPU_HAS_SYSTICK)
        __start_swap_tsc = (u32_t)SysTick->VAL;
+#elif defined(CONFIG_NRF_RTC_TIMER)
+       __start_swap_tsc = NRF_RTC1->COUNTER;
+#endif
But still i was able to see the issue on nrf52.





However this hang issue goes off, if we avoid calling function "read_systick_end_of_isr" just before calling ISR. (this experiment is with MMU disabled)
--- a/arch/arm/core/isr_wrapper.S
+++ b/arch/arm/core/isr_wrapper.S
@@ -114,7 +114,7 @@ _idle_state_cleared:
        push {r3}
 #endif
        push {lr}
-       bl read_systick_end_of_isr
+//     bl read_systick_end_of_isr
 #if defined(CONFIG_ARMV6_M)
        pop {r3}
        mov lr,r3

cvinayak commented:
@youvedeep is the call to read_systick_end_of_isr not enclosed in CONFIG_CPU_HAS_SYSTICK? if not, its clearly a defect.

youvedeep commented:
Hi @cvinayak 

For nrf based platform CONFIG_CPU_HAS_SYSTICK is not set.

Also I have changed my function "read_systick_end_of_isr" for nrf to :-
 void read_systick_end_of_isr(void)
 {
/+#if defined(CONFIG_CPU_HAS_SYSTICK)
        __end_intr_tsc  = (u32_t)SysTick->VAL;
/+#elif defined(CONFIG_NRF_RTC_TIMER)
/+       /* __end_intr_tsc = NRF_RTC1->COUNTER; */
/+       __end_tick_tsc= 123123;
/+#endif
 }

carlescufi commented:
@youvedeep @cvinayak 

Since removing the code from read_systick_end_of_isr seems to fix it, I tried not removing the code but pushing the registers that GCC is using inside the C function:

```
@@ -113,14 +113,14 @@ _idle_state_cleared:
 #if defined(CONFIG_ARMV6_M)
        push {r3}
 #endif
-       push {lr}
+       push {lr, r2, r3}
        bl read_systick_end_of_isr
 #if defined(CONFIG_ARMV6_M)
        pop {r3}
        mov lr,r3
        pop {r3}
 #else
-       pop {lr}
+       pop {r3, r2, lr}
 #endif
 #endif

```

This also seems to fix the issue, which makes me think there might be a problem with isr_wrapper.S and register saving

carlescufi commented:
@youvedeep are you on IRC by any chance? it would be great to be able to discuss this with you on a channel. If you are, just ping me, my nick is "carlesc"

youvedeep commented:
I further did 2 Experiments which states that there is something crazy going on with "reg r3" on nrf42:- 

**Experiment 1:-
Collect the value of ISR into reg r4 and move the value from r4 to r3 after benchmarking test is done :-** 
_[ysingh@basotc001 zephyr-1]$ git diff arch/arm/core/isr_wrapper.S
diff --git a/arch/arm/core/isr_wrapper.S b/arch/arm/core/isr_wrapper.S
index b26c411..949b7f5 100644
\--- a/arch/arm/core/isr_wrapper.S
\+++ b/arch/arm/core/isr_wrapper.S
@@ -108,7 +108,8 @@ _idle_state_cleared:
        add r1, r1, r0  /* table entry: ISRs must have their MSB set to stay
                         * in thumb mode */
 
\-       ldm r1!,{r0,r3} /* arg in r0, ISR in r3 */
\+       ldm r1!,{r0,r4} /* arg in r0, ISR in r3 */
\+
 #ifdef CONFIG_EXECUTION_BENCHMARKING
 #if defined(CONFIG_ARMV6_M)
        push {r3}
@@ -123,6 +124,7 @@ _idle_state_cleared:
        pop {lr}
 #endif
 #endif
\+       mov r3, r4;
        blx r3          /* call ISR */
 
 #if defined(CONFIG_ARMV6_M)_

Result:- most of cases i got test pass but a couple of times i got it failed in-between.
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL
starting test - Time Measurement
Timing Results: Clock Frequency: 0 MHz
Context switch                               : 398 cycles , 12171253 ns
Thread Creation                              :   1 cycles , 30581 ns
Thread Resume                                :   1 cycles , 30581 ns
Message Queue Put with context switch        :   1 cycles , 30581 ns
Message Queue get without context switch     :   1 cycles , 30581 ns
MailBox synchronous get                      :   1 cycles , 30581 ns
MailBox get without context switch           :   1 cycles , 30581 ns
Timing Measurement  finished
PASS - main.
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL
starting test - Time Measurement
Timing Results: Clock Frequency: 0 MHz
Context switch                               : 399 cycles , 12201834 ns
Thread abort                                 :   1 cycles , 30581 ns
Thread Resume                                :   1 cycles , 30581 ns
***** USAGE FAULT *****
  Executing thread ID (thread): 0x200011c0
  Faulting instruction address:  0x20001220
  Illegal load of EXC_RETURN into PC
Fatal fault in thread 0x200011c0! Aborting.
***** HARD FAULT *****
  Fault escalation (see below)
Fatal fault in ISR! Spinning...
starting test - Time Measurement
Timing Results: Clock Frequency: 0 MHz
Context switch                               : 398 cycles , 12171253 ns
Thread Resume                                :   1 cycles , 30581 ns
***** USAGE FAULT *****
  Executing thread ID (thread): 0x200011c0
  Faulting instruction address:  0x20001220
  Illegal load of EXC_RETURN into PC
Fatal fault in thread 0x200011c0! Aborting.
***** HARD FAULT *****
  Fault escalation (see below)
Fatal fault in ISR! Spinning...
starting test - Time Measurement
Timing Results: Clock Frequency: 0 MHz
Context switch                               : 400 cycles , 12232415 ns
Thread abort                                 :   1 cycles , 30581 ns
Thread Suspend                               :   1 cycles , 30581 ns
Thread Sleep                                 :   1 cycles , 30581 ns
Semaphore Take with context switch           :   1 cycles , 30581 ns
Message Queue Put with context switch        :   1 cycles , 30581 ns
Message Queue get without context switch     :   1 cycles , 30581 ns
MailBox synchronous get                      :   1 cycles , 30581 ns
MailBox get without context switch           :   1 cycles , 30581 ns
Timing Measurement  finished
PASS - main.
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL



**Experiment 2:
Save value of r3 into stack and pull it back after benchmarking function (read_systick_end_of_isr) is called. (somehting like patch taken by ARMV6_M).**
_[ysingh@basotc001 zephyr-1]$ git diff arch/arm/core/isr_wrapper.S
diff --git a/arch/arm/core/isr_wrapper.S b/arch/arm/core/isr_wrapper.S
index b26c411..f8718ae 100644
\--- a/arch/arm/core/isr_wrapper.S
\+++ b/arch/arm/core/isr_wrapper.S
@@ -109,19 +109,20 @@ _idle_state_cleared:
                         * in thumb mode */
 
        ldm r1!,{r0,r3} /* arg in r0, ISR in r3 */

 #ifdef CONFIG_EXECUTION_BENCHMARKING
\-#if defined(CONFIG_ARMV6_M)
\+//#if defined(CONFIG_ARMV6_M)
        push {r3}
\-#endif
\+//#endif
        push {lr}
        bl read_systick_end_of_isr
\-#if defined(CONFIG_ARMV6_M)
\+//#if defined(CONFIG_ARMV6_M)
        pop {r3}
        mov lr,r3
        pop {r3}
\-#else
\-       pop {lr}
\-#endif
\+//#else
\+//     pop {lr}
\+//#endif
 #endif
        blx r3          /* call ISR */
 
[ysingh@basotc001 zephyr-1]$_

Result :- 
I test 100 times and it was passing every time.

starting test - Time Measurement
Timing Results: Clock Frequency: 0 MHz
Context switch1                              : 400 cycles , 12232415 ns
Thread abort                                 :   1 cycles , 30581 ns
Thread Suspend                               :   1 cycles , 30581 ns
starting test - Time Measurement
Timing Results: Clock Frequency: 0 MHz
Context switch1                              : 416 cycles , 12721712 ns
Thread abort                                 :   1 cycles , 30581 ns
Thread Suspend                               :   1 cycles , 30581 ns
Thread Sleep                                 :   1 cycles , 30581 ns
Message Queue Put with context switch        :   1 cycles , 30581 ns
Message Queue get without context switch     :   1 cycles , 30581 ns
MailBox synchronous get                      :   1 cycles , 30581 ns
MailBox get without context switch           :   1 cycles , 30581 ns
Timing Measurement  finished
PASS - main.
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL





In summery, I got success in test if we workaround with r3 contents, not sure if value of "r3 is getting corrupted when function call is made just before ISR".

@galak/@carlescufi Is there a possibility that value of r3 can corrupted if call a function just before calling a ISR.
Also not sure why this issue is occurring only with nrf52, we do not see this issue on frdm board (which also takes path as taken on nrf52 (CONFIG_ARMV7_M is enabled for both).


For my experiment i used attached patch ( Experiment 1).
For experiment 2 just make the necessary changes as shown above.

carlescufi commented:
> In summery, I got success in test if we workaround with r3 contents, not sure if value of "r3 is getting corrupted when function call is made just before ISR".

@youvedeep It is definitely getting "corrupted" since GCC uses it inside the function:

```
0000065c <read_systick_end_of_isr>:
     65c:       4b02            ldr     r3, [pc, #8]    ; (668 <read_systick_end_of_isr+0xc>)
     65e:       f240 1223       movw    r2, #291        ; 0x123
     662:       601a            str     r2, [r3, #0]
     664:       4770            bx      lr
     666:       bf00            nop
     668:       20001528        .word   0x20001528
```

Not sure why it doesn't happen with the other ARM board, you could try objdump -d on the .elf and look to see if maybe GCC is not using r3 in that case

carlescufi commented:
> I feel the issue is with the AAPCS for ARM. if the scratch registers are saved it should be okay.

@andrewboie @AdithyaBaglody @youvedeep @cvinayak We think that the compiler sees the read_systick_end_of_isr being called only from an interrupt handler and therefore does not save the R0-R3 scratch registers. Hence, it should be the caller of read_systick_end_of_isr (isr_wrapper) that should save R3 and restore it after calling read_systick_end_of_isr.

youvedeep commented:
Hi @carlescufi, 
we also root caused this to saving of r0-r4 regs,  will submit correct patch.

carlescufi commented:
@andrewboie @youvedeep thanks! will you also submit a patch to adapt this test to non-systick platforms? I think we should start by renaming the `_systick_` functions to something like `_systimer_` and then adding conditional compilation of the reading of the timer, since for some ARM platforms we don't use the SysTick as a timer and the benchmarks would not be correct. Also, the MHz print in the beginning of the test is incorrect (shows 0) because we use a 32KHz timer and you divide by 1M.

youvedeep-singh committed:
arch: arm: store/restore regs r0-r3 while making function call.

In benchmark test (test_info) while making function call regs
r0 - r4 are modified into called function. Due to this value
inside r3 is getting lost.
This patch saves and restore the value in r0-r4 regs while making
function calls from assembly language.

Jira: ZEP-2314

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

nashif commented:
with this patch we might not be getting any faults anymore, but the benchmark data is bogus:

```
starting test - Time Measurement
Timing Results: Clock Frequency: 0 MHz
Context switch                               :1908 cycles , 58348623 ns
Thread Creation                              :   1 cycles , 30581 ns
Thread Resume                                :   1 cycles , 30581 ns
starting test - Time Measurement
Timing Results: Clock Frequency: 0 MHz
Context switch                               : 420 cycles , 12844036 ns
Thread Creation                              :   1 cycles , 30581 ns
starting test - Time Measurement
Timing Results: Clock Frequency: 0 MHz
Context switch                               : 404 cycles , 12354740 ns
Thread cancel                                :   1 cycles , 30581 ns
Semaphore Take with context switch           :   1 cycles , 30581 ns
Semaphore Give without context switch        :   1 cycles , 30581 ns
Message Queue Put with context switch        :   1 cycles , 30581 ns
Message Queue get without context switch     :   1 cycles , 30581 ns
MailBox synchronous get                      :   1 cycles , 30581 ns
Timing Measurement  finished
PASS - main.
===================================================================
===================================================================
PROJECT EXECUTION SUCCESSFUL

```

So what else is wrong?

carlescufi commented:
@nashif the code uses the SysTick but we don't implement it properly in hardware. The benchmarking code should use a generic timer instead, which will be SysTick for some ARM SoCs but another timer for others, like RTC for Nordic. See my comment here: https://github.com/zephyrproject-rtos/zephyr/pull/1095#issuecomment-324275000

youvedeep commented:
Hi @carlescufi , In the latest patch we are reading value through RTC Timer in case of Nordic platform (not though systick).




Hi @nashif 

Yes number looks too high with Nordic, we need to investigate on this. Number seems way too high specially in case of "Context switch " (which looks like wrong).


Here are few deltas between ARM and nrf which may contribute some change :- 
1. Some of other test results are not shown where timing is less than 1 cycle. In case of Nordic RTC timer clock speed is 32Khz (where as in case of frdm_64 it is 120MHz). 
2. by default CONFIG_SYS_CLOCK_TICKS_PER_SEC is 100 in case of nrf (Vs 1000 in case of frdm).

But still numbers are too high we need to investigate on this tomorrow.

galak commented on kernel/init.c:
```diff
@@ -67,10 +67,10 @@ u64_t __noinit __idle_time_stamp;  /* timestamp when CPU goes idle */
+u64_t __noinit __end_swap_time;
+u64_t __noinit __start_intr_time;
+u64_t __noinit __end_intr_time;
 u64_t __noinit __start_tick_tsc;
```
what about these 2?

carlescufi replied:
agreed, @youvedeep not sure what tsc stood for, but it seems like you forgot to rename a couple

youvedeep replied:
hi @carlescufi / @galak 

Yes i missed this while changing the name of other variables used. I will include this in my next version of patch.  Also as 32KHz is too slow to get the benchmark data, So we need to use timer0 for this with prescale value of 0(16MHz).

I will put new patch for review.

youvedeep commented:
Submitted new version of patch -
1. Fixed review comments.
2. Used system timer (timer2) @ 16MHz to collect stats. (as RTC 32KHz is too slow to read OS parameters).

carlescufi reviewed (dismissed):
LGTM

galak reviewed (changes requested):
I'd ask we split the 'tests: benchmark: Add support for nrf series processors.' commit so there is one commit that just does the API renaming.  Its difficult to see the other changes that are mixed in the renaming.

galak reviewed (changes requested):
I would ask we also address #1236 as part of this PR.

nashif commented:
> I would ask we also address #1236 as part of this PR.

This needs a complete redesign of how those numbers are captured and would require more time. I would like to see this change go into 1.9 to avoid having bogus numbers for NRF

galak commented:
@nashif ok, wasn't sure if we wanted this in 1.9, than I'd ask to at least make the change I asked about splitting out the rename so I can actually review the NRF changes.  I feel like this test shows either issues with our APIs that need to get logged as a bug/feature enhancement, but can't exactly tell right now.

youvedeep-singh committed:
tests: benchmark: timing_info: Change API/variable Name.

The API/Variable names in timing_info looks very speicific to
platform (like systick etc), whereas these variabled are used
across platforms (nrf/arm/quark).
So this patch :-
1. changing API/Variable names to generic one.
2. Creating some of Macros whose implimentation is platform
depenent.

Jira: ZEP-2314

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep-singh committed:
tests: benchmark: timing_info: Add support for nrf series processors

nrf SOC uses nrf rtc timer (not sys tick), which is 32kHz,
whereas CPU runs at higher speed (nrf52 runs at 64MHz).
So 32Khz is too slow to measure critical kernel parameters.

This patch does :-
1. Add support for nrf SOC for timing_info benchmarking.
2. Uses SOC timer to measure kernel parameters.

Jira: ZEP-2314

Signed-off-by: Youvedeep Singh <youvedeep.singh@intel.com>

youvedeep commented:
@galak i have split the patch into 3 parts based on your comments, please provide the review comments.

galak commented on tests/benchmarks/timing_info/src/timing_info.h:
```diff
@@ -27,10 +27,53 @@ extern u64_t __end_intr_time;
 extern u64_t __end_tick_time;
 
+/* NRF RTC TIMER runs ar very slow rate (32KHz), So in order to measure
+ * Kernel stats dedicated timer is used to measure kernel stats
```
Why do we use TIMER2?

galak reviewed (changes requested):
Seems like we should implement a counter driver for NRF and use the counter API instead of having an NRF specific version.  However, I'd be ok with doing that as a follow up cleanup to this series.

Some minor changes in #ifdef usage and comments.

youvedeep commented:
addressed review comments.

carlescufi commented:
We are working on a counter driver for the RTC, and later one for the TIMER peripheral will come as well. In the meantime, this will do the job

carlescufi commented:
Apologies, I closed by mistake

nashif commented:
fixes #1218
